IndexOutOfBoundsException during processing values with double quotes

If for example a following JSON string gets processed:

```
{
  "entries": [
    {
      "type": "internal",
      "position": "test issue \"double quote\" testing ",
      "key": "World"
    }
  ]
}
```

it fails with an error message similar to this:

> net.sf.jetro.stream.visitor.JsonIOException: An exception occurred while processing a JSON stream on line 1, column 839
> 
> Caused by: net.sf.jetro.stream.visitor.JsonIOException: An exception occurred while processing a JSON stream on line 1, column 839
>   at net.sf.jetro.stream.visitor.StreamVisitingReader.accept(StreamVisitingReader.java:38)
> 
> Caused by: java.lang.IndexOutOfBoundsException: start 52, end 17, s.length() 69
>   at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:476)
>   at java.lang.StringBuilder.append(StringBuilder.java:191)
>   at java.lang.StringBuilder.append(StringBuilder.java:72)
>   at net.sf.jetro.stream.JsonGenerator.string(JsonGenerator.java:559)
>   at net.sf.jetro.stream.JsonGenerator.value(JsonGenerator.java:418)
>   at net.sf.jetro.stream.visitor.AbstractJsonWritingVisitor.visitValue(AbstractJsonWritingVisitor.java:65)
>   at net.sf.jetro.visitor.chained.ChainedJsonVisitor.visitValue(ChainedJsonVisitor.java:112)
>   at net.sf.jetro.stream.visitor.StreamVisitingReader.acceptInternal(StreamVisitingReader.java:76)
>   at net.sf.jetro.stream.visitor.StreamVisitingReader.accept(StreamVisitingReader.java:34)
>   ... 140 more

This can be reproduced for example with a following unit test for the JsonGenerator class:

``` java
    @Test
    public void testStringWithDouleQuotesInValue() throws IOException {

        // Given:
        String valueIn = "Test msg \"double quotes\" testing";
        String expectedJsonValue = "Test msg \\\"double quotes\\\" testing";
        String expectedJson = "{\"key\":\"" + expectedJsonValue + "\"}";

        // When:
        StringBuffer out = new StringBuffer();
        JsonGenerator testee = new JsonGenerator(out);

        testee.beginObject();
        testee.name("key");
        testee.value(valueIn);
        testee.endObject();

        // Then:
        assertEquals(out.toString(), expectedJson);
    }
```

--- a/jetro-stream/src/main/java/net/sf/jetro/stream/JsonGenerator.java
+++ b/jetro-stream/src/main/java/net/sf/jetro/stream/JsonGenerator.java
@@ -550,13 +550,13 @@ public class JsonGenerator implements Closeable, Flushable {
 				continue;
 			}
 			if (last < i) {
-				out.append(value, last, i - last);
+				out.append(value, last, i);
 			}
 			out.append(replacement);
 			last = i + 1;
 		}
 		if (last < length) {
-			out.append(value, last, length - last);
+			out.append(value, last, length);
 		}
 		out.append("\"");
 	}

--- a/jetro-stream/src/test/java/net/sf/jetro/stream/visitor/JsonGeneratorTest.java
+++ b/jetro-stream/src/test/java/net/sf/jetro/stream/visitor/JsonGeneratorTest.java
@@ -32,4 +32,46 @@ public class JsonGeneratorTest {
 		// Then:
 		assertEquals(out.toString(), expectedJson);
 	}
+
+	@Test
+	public void testStringWithDouleQuotesInValueAtEnd() throws IOException {
+
+		// Given:
+		String valueIn = "Test msg \"double quotes\"";
+		String expectedJsonValue = "Test msg \\\"double quotes\\\"";
+		String expectedJson = "{\"key\":\"" + expectedJsonValue + "\"}";
+
+		// When:
+		StringBuffer out = new StringBuffer();
+		JsonGenerator testee = new JsonGenerator(out);
+
+		testee.beginObject();
+		testee.name("key");
+		testee.value(valueIn);
+		testee.endObject();
+
+		// Then:
+		assertEquals(out.toString(), expectedJson);
+	}
+
+	@Test
+	public void testStringWithDouleQuotesInValueAtBeginning() throws IOException {
+
+		// Given:
+		String valueIn = "\"double quotes\" testing";
+		String expectedJsonValue = "\\\"double quotes\\\" testing";
+		String expectedJson = "{\"key\":\"" + expectedJsonValue + "\"}";
+
+		// When:
+		StringBuffer out = new StringBuffer();
+		JsonGenerator testee = new JsonGenerator(out);
+
+		testee.beginObject();
+		testee.name("key");
+		testee.value(valueIn);
+		testee.endObject();
+
+		// Then:
+		assertEquals(out.toString(), expectedJson);
+	}
 }

--- a/jetro-stream/src/test/java/net/sf/jetro/stream/visitor/JsonGeneratorTest.java
+++ b/jetro-stream/src/test/java/net/sf/jetro/stream/visitor/JsonGeneratorTest.java
@@ -1,0 +1,35 @@
+package net.sf.jetro.stream.visitor;
+
+import net.sf.jetro.stream.JsonGenerator;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+import static org.testng.Assert.assertEquals;
+
+/**
+ * @author oliver.burkhalter
+ */
+public class JsonGeneratorTest {
+
+	@Test
+	public void testStringWithDouleQuotesInValue() throws IOException {
+
+		// Given:
+		String valueIn = "Test msg \"double quotes\" testing";
+		String expectedJsonValue = "Test msg \\\"double quotes\\\" testing";
+		String expectedJson = "{\"key\":\"" + expectedJsonValue + "\"}";
+
+		// When:
+		StringBuffer out = new StringBuffer();
+		JsonGenerator testee = new JsonGenerator(out);
+
+		testee.beginObject();
+		testee.name("key");
+		testee.value(valueIn);
+		testee.endObject();
+
+		// Then:
+		assertEquals(out.toString(), expectedJson);
+	}
+}

